### ISSUE TITLE
DTensor sharding propagation for `scaled_dot_product_efficient_attention` should be more conservatively cached

### ISSUE THREAD

**Issue by speediedan** [pytorch/pytorch#134050]
### 🐛 Describe the bug

The current `DTensor` sharding propagation caching policy for  `aten.scaled_dot_product_efficient_attention` (default) can result in silently incorrect gradients or trigger an IMA after cuda kernel launch (bt below) in mixed `require_grad` configurations.

Testing a variety of TP `requires_grad` patterns (validating mechanistic interpretability probing use cases among others) revealed that if the first `aten.scaled_dot_product_efficient_attention` sharding propagation has `compute_log_sumexp=False` (e.g. first Transformer block attention layer has `require_grad=False`), subsequent `aten.scaled_dot_product_efficient_attention` ops may re-use the sharding propagation cache inappropriately (e.g., for any later Transformer block attention layer that may have `requires_grad=True`) resulting in the aforementioned silent gradient incorrectness or IMA.

For instance, the silent correctness scenario can be observed with the test referenced below using a simple two-device mesh and the default `n_heads=4` Transformer config. It occurs because: 
- The corrupted cache triggers an extra sharding operation. `scaled_dot_product_efficient_attention_strategy` sets the `logsumexp` `DTensorSpec` to `Replicate()` and shape `[bs, n_heads, 0]`, triggering an unintended redistribute and `_replicate_to_shard` before the local backward op.
- Using the `num_heads_dim_sharding` strategy, the shard placement `mesh_dim` equals `sqrt(n_heads)` so the extra sharding operation yields a `local_tensor_args` tensor that will not will not raise an IMA when passed to our cuda kernel 

  ```cpp 
  (gdb) s
  operator()<PyTorchMemEffAttention::AttentionBackwardKernel<cutlass::arch::Sm75, float, true, false, false, 64, 64, 32>, void (*)(PyTorchMemEffAttention::AttentionBackwardKernel<cutlass::arch::Sm75, float, true, false, false, 64, 64, 32>::Params)> (__closure=0x7f369ffce4c0, _k=..., kernel_fn=0x7f37c0a58082 <fmha_cutlassB_f32_aligned_64x64_k32_sm75(PyTorchMemEffAttention::AttentionBackwardKernel<cutlass::arch::Sm75, float, true, false, false, 64, 64, 32, false>::Params)>) at /home/speediedan/repos/pytorch/aten/src/ATen/native/transformers/cuda/attention_backward.cu:693
  (gdb) p logsumexp.impl_.target_->sizes()[0]
  $6 = (const long &) @0x7f3698262048: 8
  (gdb) p logsumexp.impl_.target_->sizes()[1]
  $7 = (const long &) @0x7f3698262050: 1
  (gdb) p logsumexp.impl_.target_->sizes()[2]
  $8 = (const long &) @0x7f3698262058: 32
  (gdb) p logsumexp.impl_.target_->numel()
  $9 = 256
  ```

Alternatively, if `mesh_dim < sqrt(n_heads)` in the above scenario, when we do the unintended extra sharding, we'll pass an empty `logsumexp` tensor to one of the kernels 
```cpp

(gdb) p logsumexp.impl_.target_->sizes()[0]
$13 = (const long &) @0x7fa685058c48: 8
(gdb) p logsumexp.impl_.target_->sizes()[1]
$14 = (const long &) @0x7fa685058c50: 0
(gdb) p logsumexp.impl_.target_->sizes()[2]
$15 = (const long &) @0x7fa685058c58: 32
(gdb) p logsumexp.impl_.target_->numel()
$17 = 0
```
and subsequently trigger the IMA (w/ `PYTORCH_NO_CUDA_MEMORY_CACHING=1` and `CUDA_LAUNCH_BLOCKING=1` of course):

  ```cpp
  (gdb) bt
  #0  c10::cuda::c10_cuda_check_implementation (err=700, filename=0x7f7d6fece3a0 "/home/speediedan/repos/pytorch/c10/cuda/CUDACachingAllocator.cpp", 
      function_name=0x7f7d6fed1af7 "uncached_delete", line_number=3142, include_device_assertions=true) at /home/speediedan/repos/pytorch/c10/cuda/CUDAException.cpp:43
  #1  0x00007f7d6ff37fbe in c10::cuda::CUDACachingAllocator::Native::uncached_delete (ptr=0x7f7c75e12400) at /home/speediedan/repos/pytorch/c10/cuda/CUDACachingAllocator.cpp:3142
  #2  0x00007f7d6ea757d8 in std::unique_ptr<void, void (*)(void*)>::~unique_ptr (this=0x7f7c0ae187b8, __in_chrg=<optimized out>) at /usr/include/c++/11/bits/unique_ptr.h:361
  #3  0x00007f7d6ea6f596 in c10::detail::UniqueVoidPtr::~UniqueVoidPtr (this=0x7f7c0ae187b0, __in_chrg=<optimized out>) at /home/speediedan/repos/pytorch/c10/util/UniqueVoidPtr.h:41
  #4  0x00007f7d6ea6f62e in c10::DataPtr::~DataPtr (this=0x7f7c0ae187b0, __in_chrg=<optimized out>) at /home/speediedan/repos/pytorch/c10/core/Allocator.h:26
  #5  0x00007f7d6ea7babc in c10::StorageImpl::~StorageImpl (this=0x7f7c0ae187a0, __in_chrg=<optimized out>) at /home/speediedan/repos/pytorch/c10/core/StorageImpl.h:87
  #6  0x00007f7d6ea7bae8 in c10::StorageImpl::~StorageImpl (this=0x7f7c0ae187a0, __in_chrg=<optimized out>) at /home/speediedan/repos/pytorch/c10/core/StorageImpl.h:87
  #7  0x00007f7d6fb2b87d in c10::intrusive_ptr<c10::StorageImpl, c10::detail::intrusive_target_default_null_type<c10::StorageImpl> >::reset_ (this=0x7f7c08025560)
      at /home/speediedan/repos/pytorch/c10/util/intrusive_ptr.h:297
  #8  0x00007f7d6fb2b3d0 in c10::intrusive_ptr<c10::StorageImpl, c10::detail::intrusive_target_default_null_type<c10::StorageImpl> >::~intrusive_ptr (this=0x7f7c08025560, 
      __in_chrg=<optimized out>) at /home/speediedan/repos/pytorch/c10/util/intrusive_ptr.h:377
  #9  0x00007f7d6fb5184a in c10::Storage::~Storage (this=0x7f7c08025560, __in_chrg=<optimized out>) at /home/speediedan/repos/pytorch/c10/core/Storage.h:25
  #10 0x00007f7d6fb4b4f4 in c10::TensorImpl::~TensorImpl (this=0x7f7c08025550, __in_chrg=<optimized out>) at /home/speediedan/repos/pytorch/c10/core/TensorImpl.cpp:77
  #11 0x00007f7d6fb4b520 in c10::TensorImpl::~TensorImpl (this=0x7f7c08025550, __in_chrg=<optimized out>) at /home/speediedan/repos/pytorch/c10/core/TensorImpl.cpp:77
  #12 0x00007f7d48b76629 in c10::intrusive_ptr<c10::TensorImpl, c10::UndefinedTensorImpl>::reset_ (this=0x7f7c152feb70) at /home/speediedan/repos/pytorch/c10/util/intrusive_ptr.h:297
  #13 0x00007f7d48b73fec in c10::intrusive_ptr<c10::TensorImpl, c10::UndefinedTensorImpl>::~intrusive_ptr (this=0x7f7c152feb70, __in_chrg=<optimized out>)
      at /home/speediedan/repos/pytorch/c10/util/intrusive_ptr.h:377
  #14 0x00007f7d48b7033c in at::TensorBase::~TensorBase (this=0x7f7c152feb70, __in_chrg=<optimized out>) at /home/speediedan/repos/pytorch/aten/src/ATen/core/TensorBase.h:82
  #15 0x00007f7d48b7042e in at::Tensor::~Tensor (this=0x7f7c152feb70, __in_chrg=<optimized out>) at /home/speediedan/repos/pytorch/build/aten/src/ATen/core/TensorBody.h:92
  #16 0x00007f7d4c24b89c in operator()<PyTorchMemEffAttention::AttentionBackwardKernel<cutlass::arch::Sm75, float, true, true, false, 64, 64, 128>, void (*)(PyTorchMemEffAttention::AttentionBackwardKernel<cutlass::arch::Sm75, float, true, true, false, 64, 64, 128>::Params)> (__closure=0x7f7c152feea0, _k=..., 
      kernel_fn=0x7f7d4c6575d7 <fmha_cutlassB_f32_aligned_64x64_k128_dropout_sm75(PyTorchMemEffAttention::AttentionBackwardKernel<cutlass::arch::Sm75, float, true, true, false, 64, 64, 128, false>::Params)>) at /home/speediedan/repos/pytorch/aten/src/ATen/native/transformers/cuda/attention_backward.cu:693
  ```

The issue can be reproduced by adding a simple parameterization to the below test included in [this currently under review PR](https://github.com/pytorch/pytorch/pull/133502)

```bash
python test/distributed/tensor/parallel/test_tp_examples.py DistTensorParallelExampleTest::test_transformer_req_grad_seq_parallel_float32_thaw_layers_0_feed_forward_w2
```

```python
diff --git a/test/distributed/tensor/parallel/test_tp_examples.py b/test/distributed/tensor/parallel/test_tp_examples.py
index 82d6b2fb009..bfa308d3353 100644
--- a/test/distributed/tensor/parallel/test_tp_examples.py
+++ b/test/distributed/tensor/parallel/test_tp_examples.py
@@ -409,6 +409,17 @@ class DistTensorParallelExampleTest(DTensorTestBase):
                     bwd={reduce_scatter: 5, all_gather: 5}, optim={all_reduce: 3}
                 ),
             ),
+            (
+            (
+                    r"layers.0.feed_forward.w2.weight",
+                    r"layers.0.feed_forward.w2.bias",
+                ),  # tests sdpa op shard propagation caching logic
+                True,
+                torch.float32,
+                ExpCommCounts(
+                    bwd={reduce_scatter: 3, all_gather: 3}
+                ),
```

While there are a number of ways this could be addressed, I think the most straightforward remediation would be to modify the sharding propagation caching policy of [`aten._scaled_dot_product_efficient_attention.default`](https://github.com/pytorch/pytorch/blob/b03381cac28ec118f710eb3e808a3c1a6391abe2/torch/distributed/_tensor/ops/_matrix_ops.py#L337-L340), registering it with `schema_info=RuntimeSchemaInfo(4)` to prevent cache sharing between differing `compute_log_sumexp` values i.e.

```python
@register_op_strategy(aten._scaled_dot_product_efficient_attention.default, schema_info=RuntimeSchemaInfo(4))
def scaled_dot_product_efficient_attention_strategy(
...
```

I have a PR with the above modification and the referenced additional `test_tp_examples.py DistTensorParallelExampleTest::test_transformer_req_grad` test ready to go. As it depends on merging [#133502](https://github.com/pytorch/pytorch/pull/133502) (I don't think I have the necessary permissions to submit a ghstack), I can either:
1. add another commit to #133502 and use it to address this issue in addition to #133499 or
2. submit a separate PR if that's preferred

Just let me know whichever approach the team favors. 

Thanks again for your immensely valuable contributions to the open-source ML ecosystem, it's a real joy using and learning from your shared work!

> [!NOTE]
> Though it's lower impact and I haven't tested it, I think it may be worth looking at `aten::_scaled_dot_product_flash_attention` as well as it could be affected by a similar issue wrt `return_debug_mask`.  
>```cpp
>aten::_scaled_dot_product_flash_attention(Tensor query, Tensor key, Tensor value, float dropout_p=0.0, bool is_causal=False, >bool return_debug_mask=False, *, float? scale=None) -> (Tensor output, Tensor logsumexp, Tensor cum_seq_q, Tensor ?>cum_seq_k, SymInt max_q, SymInt max_k, Tensor philox_seed, Tensor philox_offset, Tensor debug_attn_mask)
> ```

### Versions

Collecting environment information...
PyTorch version: 2.5.0.dev20240820
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.12.4 | packaged by Anaconda, Inc. | (main, Jun 18 2024, 15:12:24) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-1013-gcp-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 4090
GPU 1: NVIDIA GeForce RTX 2070 SUPER

Nvidia driver version: 550.90.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             24
On-line CPU(s) list:                0-23
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen Threadripper 1920X 12-Core Processor
CPU family:                         23
Model:                              1
Thread(s) per core:                 2
Core(s) per socket:                 12
Socket(s):                          1
Stepping:                           1
Frequency boost:                    enabled
CPU max MHz:                        3500.0000
CPU min MHz:                        2200.0000
BogoMIPS:                           6985.82
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid amd_dcm aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate ssbd ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca sev
Virtualization:                     AMD-V
L1d cache:                          384 KiB (12 instances)
L1i cache:                          768 KiB (12 instances)
L2 cache:                           6 MiB (12 instances)
L3 cache:                           32 MiB (4 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-23
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Mitigation; untrained return thunk; SMT vulnerable
Vulnerability Spec rstack overflow: Mitigation; safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] mypy==1.10.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.1.0
[pip3] optree==0.12.1
[pip3] triton==3.0.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] numpy                     2.1.0                    pypi_0    pypi
[conda] optree                    0.12.1                   pypi_0    pypi
[conda] pytorch-cuda              12.4                 hc786d27_6    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torchtriton               3.0.0+dedb7bdf33           py312    pytorch-nightly

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jbschlosser @bhosmer @cpuhrsch @erichan1 @drisspg @mikaylagawarecki @tianyu-l

**Comment by yifuwang** [comment]
cc: @fegin, @tianyu-l

**Comment by tianyu-l** [comment]
@speediedan Thank you for helping identify the bug! We appreciate the report and the proposed solution.

I agree to fix the bug we should simply adding `RuntimeSchemaInfo` for the two ops (efficient attention fwd, flash attention fwd).

I'd suggest submitting a separate PR for this issue, both because it seems to be an independent problem, and because we need more time to digest the other issue. I can help review your PR.

Since personally I believe this is a simple fix, it's up to you whether you'd like to merge without further testing, or modify the existing test cases/file for the two attentions in https://github.com/pytorch/pytorch/blob/main/test/distributed/_tensor/test_matrix_ops.py.
